File: packages/sdks/server-sdks/gatrix-dotnet-server-sdk/src/Gatrix.Server.Sdk/Evaluation/Strategies/GradualRolloutRandomStrategy.cs
namespace Gatrix.Server.Sdk.Evaluation.Strategies;

/// <summary>
/// Gradual rollout using random percentage.
/// </summary>
public class GradualRolloutRandomStrategy : IStrategy
{
    public string Name => "gradualRolloutRandom";

    public bool IsEnabled(Models.StrategyParameters? parameters, Models.EvaluationContext context, string flagName)
    {
        var pct = parameters?.Percentage ?? 0;
        if (pct >= 100) return true;
        if (pct <= 0) return false;

        return Random.Shared.Next(1, 101) <= pct;
    }
}

File: packages/evaluator/src/strategies/GradualRolloutRandomStrategy.ts
/**
 * Gradual Rollout Random Strategy
 *
 * Percentage-based rollout using random value each evaluation.
 * No stickiness - purely random on every check.
 */

import { EvaluationContext, StrategyParameters, StrategyEvaluationResult } from '@gatrix/shared';
import { Strategy } from './Strategy';

export class GradualRolloutRandomStrategy extends Strategy {
  constructor() {
    super('gradualRolloutRandom');
  }

  isEnabled(parameters: StrategyParameters, _context: EvaluationContext): boolean {
    return this.isEnabledWithDetails(parameters, _context).enabled;
  }

  isEnabledWithDetails(
    parameters: StrategyParameters,
    _context: EvaluationContext
  ): StrategyEvaluationResult {
    const percentage = Number(parameters.percentage ?? 0);
    const random = Math.floor(Math.random() * 100) + 1;
    const enabled = percentage >= random;

    return {
      enabled,
      reason: enabled
        ? `Percentage ${percentage}%: random ${random} <= ${percentage}`
        : `Percentage ${percentage}%: random ${random} > ${percentage}`,
      details: { percentage, random },
    };
  }
}
